transactor/src/main/scala/io/mediachain/transactor/CborSerialization.scala

```diff
@@ -0,0 +1,452 @@
+        ))
+    }
+
+  /**
+    * Try to deserialize some `CborSerializable` object from a byte array.
+    * @param bytes an array of (presumably) cbor-encoded data
+    * @return a `CborSerializable` object, or a `DeserializationError` on failure
+    */
+  def fromCborBytes(bytes: Array[Byte]): Xor[DeserializationError, CborSerializable] =
+    CborCodec.decode(bytes) match {
+      case (_: CTag) :: (taggedValue: CValue) :: _ => fromCbor(taggedValue)
+      case (cValue: CValue) :: _ => fromCbor(cValue)
+      case Nil => Xor.left(CborDecodingFailed())
+    }
+
+  /**
+    * Try to deserialize a `CborSerializable` object from a cbor `CValue`
+    * @param cValue the `CValue` to decode
+    * @return a `CborSerializable` object, or a `DeserializationError` on failure
+    */
+  def fromCbor(cValue: CValue): Xor[DeserializationError, CborSerializable] =
+    cValue match {
+      case (cMap: CMap) => fromCMap(cMap)
+      case _ => Xor.left(UnexpectedCborType(
+        s"Expected CBOR map, but received ${cValue.getClass.getName}"
+      ))
+    }
+
+  /**
+    * Try to deserialize some `CborSerializable` object from a cbor `CMap`.
+    * @param cMap a cbor map representing the object to decode
+    * @return a `CborSerializable` object, or a `DeserializationError` on failure
+    */
+  def fromCMap(cMap: CMap)
+  : Xor[DeserializationError, CborSerializable] = {
+
+    // TODO: create multiple deserializer maps for different contexts
+    // e.g. DataStore should deserialize chain cells to specific subtypes, etc
+    val transactorDeserializers
+    : Map[String, CborDeserializer[CborSerializable]] = Map(
+      CBORTypeNames.Entity -> EntityDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.Artefact -> ArtefactDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.EntityChainCell -> EntityChainCellDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.ArtefactChainCell -> ArtefactChainCellDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.CanonicalEntry -> CanonicalEntryDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.ChainEntry -> ChainEntryDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]],
+
+      CBORTypeNames.JournalBlock -> JournalBlockDeserializer
+        .asInstanceOf[CborDeserializer[CborSerializable]]
+    )
+
+    for {
+      typeName <- getTypeName(cMap)
+      deserializer <- Xor.fromOption(
+        transactorDeserializers.get(typeName),
+        UnknownObjectType(typeName)
+      )
+      value <- deserializer.fromCMap(cMap)
+    } yield value
+  }
+
+
+  /**
+    * String constants for the `type` field used to indicate the type of
+    * object represented by a cbor map.
+    */
+  object CBORTypeNames {
+    val Entity = "entity"
```

Comment:
oh, that's cool.  I didn't realize you could use case objects like that.  That seems much better than using strings to me.